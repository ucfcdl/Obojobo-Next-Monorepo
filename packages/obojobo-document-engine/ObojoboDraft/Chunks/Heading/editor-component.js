import React from 'react'
import isOrNot from '../../../src/scripts/common/isornot'

class Heading extends React.Component {
	constructor(props) {
		super(props)
		this.state = {
			isOpen: false
		}

		this.handleClick = this.handleClick.bind(this)
	}

	componentDidMount() {
		document.addEventListener('mousedown', this.handleClick, false)
	}

	componentWillUnmount() {
		document.removeEventListener('mousedown', this.handleClick, false)
	}

	handleClick(event) {
		if (!this.node || this.node.contains(event.target)) return

		this.handleOutsideClick()
	}

	handleLevelChange(num) {
		const editor = this.props.editor
		const change = editor.value.change()

		this.setState({ isOpen: false })

		change.setNodeByKey(this.props.node.key, {
			data: {
				content: {
					level: num,
					align: this.props.node.data.get('content').align
				}
			}
		})
		editor.onChange(change)
	}
	toggleLevelSelect() {
		this.setState(state => {
			return { isOpen: !state.isOpen }
		})
	}
	handleOutsideClick() {
		this.setState({ isOpen: false })
	}

	render() {
		const content = this.props.node.data.get('content')
		const HTag = `h${content.level || 1}`
		const text = this.props.node.text
		const dropText = text.slice(0, 15) || 'Your Heading'

		return (
			<div
				className={'text-chunk obojobo-draft--chunks--heading pad'}
				ref={node => {
					this.node = node
				}}
			>
				<HTag>
					<span className={'text align-' + content.align}>{this.props.children}</span>
				</HTag>

				<div className={'dropdown-heading'}>
					<button onClick={() => this.toggleLevelSelect()}>{'â–¼'}</button>
					<div className={'drop-content-heading ' + isOrNot(this.state.isOpen, 'open')}>
						<button onClick={() => this.handleLevelChange(1)}>
							<h1>{dropText}</h1>
						</button>
						<button onClick={() => this.handleLevelChange(2)}>
							<h2>{dropText}</h2>
						</button>
						<button onClick={() => this.handleLevelChange(3)}>
							<h3>{dropText}</h3>
						</button>
						<button onClick={() => this.handleLevelChange(4)}>
							<h4>{dropText}</h4>
						</button>
						<button onClick={() => this.handleLevelChange(5)}>
							<h5>{dropText}</h5>
						</button>
						<button onClick={() => this.handleLevelChange(6)}>
							<h6>{dropText}</h6>
						</button>
					</div>
				</div>
			</div>
		)
	}
}

export default Heading
