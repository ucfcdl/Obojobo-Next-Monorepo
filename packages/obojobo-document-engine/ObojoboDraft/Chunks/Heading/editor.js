import React from 'react'
import Common from 'Common'

import emptyNode from './empty-node.json'
import Icon from './icon'
import Node from './editor-component'
import Schema from './schema'
import Converter from './converter'

const HEADING_NODE = 'ObojoboDraft.Chunks.Heading'

const plugins = {
	renderPlaceholder(props, editor, next) {
		const { node } = props
		if (node.object !== 'block' || node.type !== HEADING_NODE) return next()
		if (node.text !== '') return next()

		return (
			<span
				className={'placeholder align-' + node.data.get('content').align}
				contentEditable={false}
			>
				{'Type Your Heading Here'}
			</span>
		)
	},
	renderNode(props, editor, next) {
		switch (props.node.type) {
			case HEADING_NODE:
				return <Node {...props} {...props.attributes} />
			default:
				return next()
		}
	},
	schema: Schema
}

Common.Store.registerEditorModel('ObojoboDraft.Chunks.Heading', {
	name: 'Heading',
	icon: Icon,
	isInsertable: true,
	insertJSON: emptyNode,
	slateToObo: Converter.slateToObo,
	oboToSlate: Converter.oboToSlate,
	plugins
})

const Heading = {
	name: HEADING_NODE,
	components: {
		Node,
		Icon
	},
	helpers: {
		slateToObo: Converter.slateToObo,
		oboToSlate: Converter.oboToSlate
	},
	json: {
		emptyNode
	},
	plugins
}

export default Heading
