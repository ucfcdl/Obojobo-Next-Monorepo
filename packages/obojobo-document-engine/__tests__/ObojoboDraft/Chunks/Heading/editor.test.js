import Heading from '../../../../ObojoboDraft/Chunks/Heading/editor'
const HEADING_NODE = 'ObojoboDraft.Chunks.Heading'

describe('Heading editor', () => {
	test('plugins.renderNode renders a button when passed', () => {
		const props = {
			attributes: { dummy: 'dummyData' },
			node: {
				type: HEADING_NODE,
				data: {
					get: () => {
						return {}
					}
				}
			}
		}

		expect(Heading.plugins.renderNode(props)).toMatchSnapshot()
	})

	test('plugins.renderPlaceholder exits when not relevent', () => {
		expect(
			Heading.plugins.renderPlaceholder({
				node: {
					object: 'text'
				}
			})
		).toMatchSnapshot()

		expect(
			Heading.plugins.renderPlaceholder({
				node: {
					object: 'block',
					type: 'mockType'
				}
			})
		).toMatchSnapshot()

		expect(
			Heading.plugins.renderPlaceholder({
				node: {
					object: 'block',
					type: HEADING_NODE,
					text: 'Some text'
				}
			})
		).toMatchSnapshot()
	})

	test('plugins.renderPlaceholder renders a placeholder', () => {
		expect(
			Heading.plugins.renderPlaceholder({
				node: {
					object: 'block',
					type: HEADING_NODE,
					text: '',
					data: { get: () => ({ align: 'left' }) }
				}
			})
		).toMatchSnapshot()
	})
})
