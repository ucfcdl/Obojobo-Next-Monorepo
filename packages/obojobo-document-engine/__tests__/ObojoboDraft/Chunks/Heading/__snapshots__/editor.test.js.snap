// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Heading editor plugins.renderNode renders a button when passed 1`] = `
<Heading
  attributes={
    Object {
      "dummy": "dummyData",
    }
  }
  dummy="dummyData"
  node={
    Object {
      "data": Object {
        "get": [Function],
      },
      "type": "ObojoboDraft.Chunks.Heading",
    }
  }
/>
`;

exports[`Heading editor plugins.renderPlaceholder exits when not relevent 1`] = `undefined`;

exports[`Heading editor plugins.renderPlaceholder exits when not relevent 2`] = `undefined`;

exports[`Heading editor plugins.renderPlaceholder exits when not relevent 3`] = `undefined`;

exports[`Heading editor plugins.renderPlaceholder renders a placeholder 1`] = `
<span
  className="placeholder align-left"
  contentEditable={false}
>
  Type Your Heading Here
</span>
`;
